diff --git a/src/miners/fs/org.freedesktop.Tracker1.Miner.Files.service.in.in b/src/miners/fs/org.freedesktop.Tracker1.Miner.Files.service.in.in
index 1a285d9..009302b 100644
--- a/src/miners/fs/org.freedesktop.Tracker1.Miner.Files.service.in.in
+++ b/src/miners/fs/org.freedesktop.Tracker1.Miner.Files.service.in.in
@@ -1,6 +1,7 @@
 [D-BUS Service]
 Name=org.freedesktop.Tracker1.Miner.Files
 Exec=@libexecdir@/tracker-miner-fs
+SystemdService=tracker-miner-fs.service
 
 # Miner details needed for tracker-control
 Path=/org/freedesktop/Tracker1/Miner/Files
diff --git a/src/tracker-extract/org.freedesktop.Tracker1.Miner.Extract.service.in.in b/src/tracker-extract/org.freedesktop.Tracker1.Miner.Extract.service.in.in
index 5c4ad14..e436faa 100644
--- a/src/tracker-extract/org.freedesktop.Tracker1.Miner.Extract.service.in.in
+++ b/src/tracker-extract/org.freedesktop.Tracker1.Miner.Extract.service.in.in
@@ -1,6 +1,7 @@
 [D-BUS Service]
 Name=org.freedesktop.Tracker1.Miner.Extract
 Exec=@libexecdir@/tracker-extract
+SystemdService=tracker-extract.service
 
 # Miner details needed for tracker-control
 Path=/org/freedesktop/Tracker1/Miner/Extract
diff --git a/src/tracker-store/org.freedesktop.Tracker1.service.in b/src/tracker-store/org.freedesktop.Tracker1.service.in
index 77edc00..d834273 100644
--- a/src/tracker-store/org.freedesktop.Tracker1.service.in
+++ b/src/tracker-store/org.freedesktop.Tracker1.service.in
@@ -1,3 +1,4 @@
 [D-BUS Service]
 Name=org.freedesktop.Tracker1
 Exec=@libexecdir@/tracker-store
+SystemdService=tracker-store.service
