From 8e6735902927e60a2db8c63d7c880e8e469539bf Fri, 16 Dec 2016 14:43:20 +0100
From: Andrew Branson <andrew.branson@jollamobile.com>
Date: Wed, 10 Aug 2016 11:53:35 +0200
Subject: [PATCH] [tracker] Fix flac tag extraction. Fixes JB#35939

diff --git a/src/tracker-extract/10-flac.rule b/src/tracker-extract/10-flac.rule
index 150301c..6b46588 100644
--- a/src/tracker-extract/10-flac.rule
+++ b/src/tracker-extract/10-flac.rule
@@ -1,4 +1,4 @@
 [ExtractorRule]
 ModulePath=libextract-flac.so
-MimeTypes=audio/x-flac
+MimeTypes=audio/flac;audio/x-flac
 FallbackRdfTypes=nmm:MusicPiece;nfo:Audio;
diff --git a/src/tracker-extract/tracker-extract-flac.c b/src/tracker-extract/tracker-extract-flac.c
index ba231b8..c38a143 100644
--- a/src/tracker-extract/tracker-extract-flac.c
+++ b/src/tracker-extract/tracker-extract-flac.c
@@ -77,16 +77,15 @@
 		FLAC__StreamMetadata_VorbisComment_Entry entry;
 
 		entry = comment->comments[i];
-
 		/* entry.entry is the format NAME=metadata */
 		if (g_ascii_strncasecmp (entry.entry, "title", 5) == 0) {
 			fd->title = g_strdup (entry.entry + 6);
-		} else if (g_ascii_strncasecmp (entry.entry, "artist", 6) == 0) {
+		} else if (g_ascii_strncasecmp (entry.entry, "artist=", 7) == 0) {
 			/* FIXME: Handle multiple instances of artist */
 			if (fd->artist == NULL) {
 				fd->artist = g_strdup (entry.entry + 7);
 			}
-		} else if (g_ascii_strncasecmp (entry.entry, "album", 5) == 0) {
+		} else if (g_ascii_strncasecmp (entry.entry, "album=", 6) == 0) {
 			fd->album = g_strdup (entry.entry + 6);
 		} else if (g_ascii_strncasecmp (entry.entry, "albumartist", 11) == 0) {
 			fd->albumartist = g_strdup (entry.entry + 12);
