From 9a472dee86d0b79f23168f318f3d8ab54c64def0 Fri, 16 Dec 2016 15:18:04 +0100
From: Andrew Branson <andrew.branson@jollamobile.com>
Date: Wed, 12 Oct 2016 09:41:45 +0200
Subject: [PATCH] [tracker] FLAC metadata iterator wasn't closed. Fixes MER#1670

diff --git a/src/tracker-extract/tracker-extract-flac.c b/src/tracker-extract/tracker-extract-flac.c
index c38a143..cda8ca6 100644
--- a/src/tracker-extract/tracker-extract-flac.c
+++ b/src/tracker-extract/tracker-extract-flac.c
@@ -225,6 +225,8 @@
 		}
 	} while (FLAC__metadata_simple_iterator_next (iter));
 
+	FLAC__metadata_simple_iterator_delete (iter);
+
 	creator = tracker_coalesce_strip (3, fd.artist, fd.albumartist,
 	                                  fd.performer);
 
